### PR TITLE
[feature](Nereids): Eliminate outer join

### PR DESCRIPTION
**Pull request by jackwener**
# Proposed changes

Issue Number: close #xxx

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [x] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [x] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [x] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/EliminateOuter.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+                    for (Expression predicate : predicates) {
+                        // TODO: more case.
+                        if (predicate instanceof ComparisonPredicate) {
+                            notNullSlots.addAll(predicate.getInputSlots());
+                        }
+                    }
+                    LogicalJoin<GroupPlan, GroupPlan> join = filter.child();
+                    Pair<List<Slot>, List<Slot>> onClauseUsedSlots = JoinUtils.getOnClauseUsedSlots(join);
+                    JoinType joinType = join.getJoinType();
+                    if (!joinType.isLeftOuterJoin() && notNullSlots.containsAll(onClauseUsedSlots.first)) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/EliminateOuter.java:75]: not only on clause, we could eliminate right outer if not null slots contains any slots of left child.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/util/JoinUtils.java [hunk 2]
```diff
@@ -185,6 +186,42 @@ public static Pair<List<ExprId>, List<ExprId>> getOnClauseUsedSlots(
         return childSlotsExprId;
     }
 
+    /**
+     * Get all used slots from onClause of join.
+     */
+    public static Pair<List<Slot>, List<Slot>> getOnClauseUsedSlots(
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/util/JoinUtils.java:192]: we have a similar function i think

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/EliminateOuter.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    public static Map<JoinType, JoinType> ELIMINATE_RIGHT_MAP = ImmutableMap.of(
+            JoinType.RIGHT_OUTER_JOIN, JoinType.INNER_JOIN,
+            JoinType.FULL_OUTER_JOIN, JoinType.LEFT_OUTER_JOIN
+    );
+
+    @Override
+    public Rule build() {
+        return logicalFilter(logicalJoin())
+                .when(filter -> filter.child().getJoinType().isOuterJoin())
+                .then(filter -> {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/EliminateOuter.java:63]: i think this rule could cause infinite loop?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/EliminateOuter.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+                    List<Expression> predicates = ExpressionUtils.extractConjunction(filter.getPredicates());
+                    Set<Slot> notNullSlots = new HashSet<>();
+                    for (Expression predicate : predicates) {
+                        // TODO: more case.
+                        if (predicate instanceof ComparisonPredicate) {
```
Comment (qzsee) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/EliminateOuter.java:66]: I think `<=>`  needs special process

Comment (qzsee) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/EliminateOuter.java:66]: The handling of null values, using constant folding, is more general and does not require writing long logical judgments

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/EliminateOuter.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+                            notNullSlots.addAll(predicate.getInputSlots());
+                        }
+                    }
+                    LogicalJoin<GroupPlan, GroupPlan> join = filter.child();
+                    JoinType joinType = join.getJoinType();
+                    if (!joinType.isLeftOuterJoin() && ExpressionUtils.isIntersecting(join.left().getOutputSet(),
```
Comment (qzsee) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/EliminateOuter.java:72]: The handling of null values, using constant folding, is more general and does not require writing long logical judgments

**Comment by jackwener** [comment]
Future: todo
use constant viariable to handle it (we can handle more case like nullsafeEqual ......)

**Review by morrySnow — APPROVED** [review]

**Pull request merged**